File: genkit-tools/common/src/types/flow.ts
/**
 * Copyright 2024 Google LLC
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

import { extendZodWithOpenApi } from '@asteasolutions/zod-to-openapi';
import * as z from 'zod';

extendZodWithOpenApi(z);

// NOTE: Keep this file in sync with genkit/flow/src/types.ts!
// Eventually tools will be source of truth for these types (by generating a
// JSON schema) but until then this file must be manually kept in sync

/**
 * The message format used by the flow task queue and control interface.
 */
export const FlowInvokeEnvelopeMessageSchema = z.object({
  // Start new flow.
  start: z
    .object({
      input: z.unknown().optional(),
      labels: z.record(z.string(), z.string()).optional(),
    })
    .optional(),
  // Schedule new flow.
  schedule: z
    .object({
      input: z.unknown().optional(),
      delay: z.number().optional(),
    })
    .optional(),
  // Run previously scheduled flow.
  runScheduled: z
    .object({
      flowId: z.string(),
    })
    .optional(),
  // Retry failed step (only if step is setup for retry)
  retry: z
    .object({
      flowId: z.string(),
    })
    .optional(),
  // Resume an interrupted flow.
  resume: z
    .object({
      flowId: z.string(),
      payload: z.unknown().optional(),
    })
    .optional(),
  // State check for a given flow ID. No side effects, can be used to check flow state.
  state: z
    .object({
      flowId: z.string(),
    })
    .optional(),
});
export type FlowInvokeEnvelopeMessage = z.infer<
  typeof FlowInvokeEnvelopeMessageSchema
>;

export const FlowActionInputSchema = FlowInvokeEnvelopeMessageSchema.extend({
  auth: z.unknown().optional(),
});

export const FlowStateExecutionSchema = z.object({
  startTime: z
    .number()
    .optional()
    .describe('start time in milliseconds since the epoch'),
  endTime: z
    .number()
    .optional()
    .describe('end time in milliseconds since the epoch'),
  traceIds: z.array(z.string()),
});
export type FlowStateExecution = z.infer<typeof FlowStateExecutionSchema>;

export const FlowResponseSchema = z.object({
  response: z.unknown().nullable(),
});
export const FlowErrorSchema = z.object({
  error: z.string().optional(),
  stacktrace: z.string().optional(),
});
export type FlowError = z.infer<typeof FlowErrorSchema>;

export const FlowResultSchema = FlowResponseSchema.and(FlowErrorSchema);

File: py/packages/genkit/src/genkit/_core/_flow.py
# Copyright 2025 Google LLC
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#     http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.
#
# SPDX-License-Identifier: Apache-2.0

"""Flow registration for Genkit."""

from __future__ import annotations

import inspect
from collections.abc import Awaitable, Callable
from typing import Any, overload

from typing_extensions import TypeVar

from genkit._core._action import Action, ActionKind, ActionRunContext, get_func_description
from genkit._core._registry import Registry

InputT = TypeVar('InputT')
OutputT = TypeVar('OutputT')


@overload
def define_flow(
    registry: Registry,
    func: Callable[[], Awaitable[OutputT]],
    name: str | None = None,
    description: str | None = None,
) -> Action[None, OutputT]: ...


@overload
def define_flow(
    registry: Registry,
    func: Callable[[InputT], Awaitable[OutputT]],
    name: str | None = None,
    description: str | None = None,
) -> Action[InputT, OutputT]: ...


@overload
def define_flow(
    registry: Registry,
    func: Callable[[InputT, ActionRunContext], Awaitable[OutputT]],
    name: str | None = None,
    description: str | None = None,
) -> Action[InputT, OutputT]: ...


def define_flow(
    registry: Registry,
    func: Callable[..., Awaitable[Any]],
    name: str | None = None,
    description: str | None = None,
) -> Action[Any, Any]:
    """Register an async function as a flow action."""
    # All Python functions have __name__, but ty is strict about Callable protocol
    if not inspect.iscoroutinefunction(func):
        raise TypeError(f'Flow must be async: {func.__name__}')  # ty: ignore[unresolved-attribute]

    flow_name = name or func.__name__  # ty: ignore[unresolved-attribute]
    return registry.register_action(
        name=flow_name,
        kind=ActionKind.FLOW,
        fn=func,
        description=get_func_description(func, description),
        span_metadata={'genkit:metadata:flow:name': flow_name},
    )
